File: C++/Algorithm/linklist/linklist.c
#include "linklist.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

LinkList* createLinkList()
{
	LinkList* head = (LinkList*)malloc(sizeof(LinkList));
	if(head == NULL){
		printf("createLinkList warn: head malloc failed\n");
		return NULL;
	}
	head->data = -1;
	head->next = NULL;
	
	return head;
}

void destroyLinkList(LinkList** linklist)
{
	if(*linklist == NULL){
		printf("destroyLinkList warn: *linklist = NULL\n");
		return ;
	}
	
	clearLinkList(*linklist);
	
	free(*linklist);
	*linklist = NULL;	
}

LinkList* initLinkList(LinkList* linklist,const int size,const data_t data)
{
	if(linklist == NULL || size <= 0){
		printf("initLinkList warn: linklist=NULL or size is illegal\n");
		return NULL;
	}
	
	clearLinkList(linklist);
	
	LinkList* p = linklist;
	for(int i=0;i<size;++i){
		LinkList* newNode = (LinkList*)malloc(sizeof(LinkList));
		if(newNode == NULL){
			printf("initLinkList warn: newNode malloc failed\n");
			clearLinkList(linklist);
			return NULL;
		}
		newNode->data = data;
		newNode->next = NULL;
		
		p->next = newNode;
		p = p->next;
	}
	
	return linklist;
}

void clearLinkList(LinkList* linklist)
{
	if(linklist == NULL){
		printf("clearLinkList warn: linklist=NULL\n");
		return ;
	}
	
	LinkList* p = linklist->next;
	LinkList* q = NULL;
	while(p != NULL){
		q = p->next;
		free(p);
		p = q;
	}
	linklist->next = NULL;
}

bool isNull(const LinkList* linklist)
{
	return linklist == NULL ? true : false;
}

bool isEmpty(const LinkList* linklist)
{
	if(linklist == NULL){
		return true;
	}
	
	return linklist->next == NULL ? true : false;
}

int getLength(const LinkList* linklist)
{
	if(linklist == NULL){
		printf("getLength warn: linklist=NULL\n");
		return -1;
	}
	
	int len = 0;
	LinkList* p = linklist->next;
	while(p != NULL){
		p = p->next;
		len++;
	}
	
	return len;
}

bool append(LinkList* linklist,const data_t data)
{
	if(isNull(linklist)){
		printf("append warn: linklist is NULL\n");
		return false;
	}
	
	LinkList* p = linklist;
	while(p->next != NULL){
		p = p->next;
	}
	
	LinkList* newNode = (LinkList*)malloc(sizeof(LinkList));
	if(newNode == NULL){
		printf("append warn: newNode malloc failed\n");
		return false;
	}
	newNode->data = data;
	newNode->next = NULL;
	
	p->next = newNode;
	return true;
}

bool prepend(LinkList* linklist,const data_t data)
{
	if(isNull(linklist)){
		printf("prepend warn: linklist is NULL\n");
		return false;
	}
	
	LinkList* p = linklist->next;
	
	LinkList* newNode = (LinkList*)malloc(sizeof(LinkList));
	if(newNode == NULL){
		printf("prepend warn: newNode malloc failed\n");
		return false;
	}
	newNode->data = data;
	newNode->next = NULL;

	linklist->next = newNode;
	linklist->next->next = p;
	return true;
}

bool insert(LinkList* linklist,const int pos,const data_t data)
{
	if(isNull(linklist) || pos < 0 || pos > getLength(linklist)){
		printf("insert warn: linklist is NULL or pos is illegal\n");
		return false;
	}
	
	LinkList* p = linklist;
	for(int i=0;i<pos;++i){
		p = p->next;
	}

	LinkList* newNode = (LinkList*)malloc(sizeof(LinkList));
	if(newNode == NULL){
		printf("insert warn: newNode malloc failed\n");
		return false;
	}
	newNode->data = data;
	newNode->next = p->next;
	
	p->next = newNode;
	return true;
}

data_t findByPos(const LinkList* linklist,const int pos)
{
	assert(!isEmpty(linklist) && pos >= 0 && pos < getLength(linklist));
	
	LinkList* p = linklist->next;
	for(int i=0;i<pos;++i){
		p = p->next;
	}
	
	return p->data;
}

int findByData_L(const LinkList* linklist,const data_t data)
{
	if(isEmpty(linklist)){
		printf("findByData_L warn: linklist is empty\n");
		return -1;
	}
	
	LinkList* p = linklist->next;
	for(int i=0;p!=NULL;++i){
		if(p->data == data){
			return i;
		}
		p = p->next;
	}
	
	return -1;
}

int findByData_R(const LinkList* linklist,const data_t data)
{
	if(isEmpty(linklist)){
		printf("findByData_L warn: linklist is empty\n");
		return -1;
	}
	
	int pos;
	LinkList* p =NULL;
	for(int i=getLength(linklist)-1;i>=0;--i){		
		pos = i;
		p = linklist->next;
		
		while(pos--){
			p = p->next;
		}
		
		if(p->data == data){
			return i;
		}
	}
	
	return -1;
}

bool deleteByPos(LinkList* linklist,const int pos)
{
	if(isEmpty(linklist) || pos < 0 || pos >= getLength(linklist)){
		printf("deleteByPos warn: linklist is empty or pos is illegal\n");
		return false;
	}
	
	LinkList* p = linklist;
	LinkList* q = NULL;
	for(int i=0;i<pos;++i){
		p = p->next;
	}
	q = p->next->next;
	
	free(p->next);
	p->next = q;
	
	return true;
}

bool deleteByData(LinkList* linklist,const data_t data)
{
	if(isEmpty(linklist)){
		printf("deleteByData warn: linklist is empty\n");
		return false;
	}
	
	int index;
	bool flag = false;
	while((index = findByData_L(linklist,data)) != -1){
		deleteByPos(linklist,index);
		flag = true;
	}
	
	return flag;
}

bool removeFirst(LinkList* linklist)
{
	if(isEmpty(linklist)){
		printf("removeFirst warn: linklist is empty\n");
		return false;
	}
	
	LinkList* p = linklist->next;
	linklist->next = p->next;
	free(p);
	p = NULL;
	
	return true;
}

bool removeLast(LinkList* linklist)
{
	if(isEmpty(linklist)){
		printf("removeFirst warn: linklist is empty\n");
		return false;
	}
	
	LinkList* p = linklist;
	while(p->next->next != NULL){
		p = p->next;
	}
	
	free(p->next);
	p->next = NULL;
	
	return true;
}

bool changeByPos(LinkList* linklist,const int pos,const data_t data)
{
	if(isEmpty(linklist) || pos < 0 || pos >= getLength(linklist)){
		printf("changeByPos warn: linklist is empty or pos is illegal\n");
		return false;
	}
	
	LinkList* p = linklist->next;
	for(int i=0;i<pos;++i){
		p = p->next;
	}
	
	p->data = data;
	return true;	
}

bool changeByData(LinkList* linklist,const data_t n,const data_t m)
{
	if(isEmpty(linklist)){
		printf("changeByData warn: linklist is empty\n");
		return false;
	}
	
	int index;
	bool flag = false;
	while((index = findByData_L(linklist,n)) != -1){
		changeByPos(linklist,index,m);
		flag = true;
	}
	
	return flag;
}

void showLinkList(const LinkList* linklist)
{
	if(isEmpty(linklist)){
		printf("showList warn: linklist is empty\n");
		return ;
	}
	
	LinkList* p = linklist->next;
	while(p != NULL){
		printf("%d ",p->data);
		p = p->next;
	}
	printf("\n");
}

File: Go/Algorithm/linklist/src/linklist/linklist.go
package linklist

import "os"
import "fmt"
import "math"

const Accuracy float64 = 0.000001

type node struct{
	data float64
	next *node
}

type LinkList struct{
	length int
	head node
}

func CreateLinkList() *LinkList{
	return new(LinkList)            //Go语言中数值类型的默认值为0，指针类型的默认值为nil
}

func (ls *LinkList)IsNull() bool{
	if ls == nil {
		return true
	}
	return false
}

func (ls *LinkList)IsEmpty() bool{
	if ls == nil || ls.head.next == nil {
		return true
	}
	
	return false
}

func (ls *LinkList)Length() int{
	if ls == nil {
		return -1
	}
	
	return ls.length
}

func (ls *LinkList)Append(value float64){
	if ls == nil {
		fmt.Println("Append() error: ls is nil")
		return
	}
	
	p := &ls.head
	for p.next != nil {
		p = p.next
	}
	
	var newNode *node = new(node)
	newNode.data = value
	newNode.next = nil                 //可以不用手动赋值，Go语言中指针类型的默认值为nil
	
	ls.length += 1
	p.next = newNode
}

func (ls *LinkList)Prepend(value float64){
	if ls == nil {
		fmt.Println("Prepend() error: ls is nil")
		return
	}
	
	var newNode *node = new(node)
	newNode.data = value
	newNode.next = ls.head.next
	
	ls.length += 1
	ls.head.next = newNode
}

func (ls *LinkList)PopFront(){
	if ls.IsEmpty() {
		return
	}
	
	var p = ls.head.next
	if p != nil {
		ls.head.next = p.next
	}
	
	ls.length--
}

func (ls *LinkList)PopBack(){
	if ls.IsEmpty() {
		return
	}
	
	p := &(ls.head)
	for p.next.next != nil {
		p = p.next
	}
	
	p.next = nil
	ls.length--
}

func (ls *LinkList)Insert(pos int,value float64){
	if ls == nil || pos < 0 || pos > ls.length {
		fmt.Println("Insert() error: ls is nil or pos is illegal")
		return
	}
	
	var p *node = &ls.head
	for ;pos > 0;pos--{
		p = p.next
	}
	
	newNode := new(node)
	newNode.data = value
	newNode.next = p.next
	
	p.next = newNode
	ls.length++
}

func (ls *LinkList)Remove(value float64){
	if ls.IsEmpty() {
		return
	}
	
	var p = &ls.head
	for p.next != nil {
		if math.Abs(p.next.data - value) < Accuracy {
			p.next = p.next.next
			ls.length--
		} else {
			p = p.next
		}
	}
}

func (ls *LinkList)At(index int) float64{
	if ls.IsEmpty() || index < 0 || index >= ls.length {
		fmt.Println("At() error: ls is empty or index is illegal")
		os.Exit(-1)
	}
	
	var p *node = ls.head.next
	for ;index > 0;index-- {
		p = p.next
	}
	
	return p.data
}

func (ls *LinkList)Contains(value float64) bool{
	if ls.IsEmpty() {
		return false
	}
	
	p := ls.head.next
	for p != nil {
		if math.Abs(p.data - value) < Accuracy {
			return true
		}
		p = p.next
	}
	
	return false
}

func (ls *LinkList)Find(value float64) int{
	if ls.IsEmpty() {
		return -1
	}
	
	p := ls.head.next
	for index := 0 ; p != nil ; index++ {
		if math.Abs(p.data - value) < Accuracy {
			return index
		}
		p = p.next
	}
	
	return -1
}

func (ls *LinkList)Change(index int,value float64){
	if ls.IsEmpty() || index < 0 || index >= ls.length {
		fmt.Println("Change() error: ls is empty or index is illegal")
		return
	}
	
	p := ls.head.next
	for ;index > 0;index-- {
		p = p.next
	}
	
	p.data = value
}

func (ls *LinkList)Display(){
	var p *node = ls.head.next
	for p != nil {
		fmt.Printf("%f ",p.data)
		p = p.next
	}
	fmt.Println("")
}